### ISSUE TITLE
[Bug] TF tracing fails for distributed training with stateful optimizers

### ISSUE THREAD

**Issue by peinguim** [keras-team/keras#23085]
Hi!

When trying to use a "stateful" optimizer (i.e. with optimizer variables, e.g. SGD with momentum, Adam) within a custom TF loop in distributed training, I keep getting these TF graph/tracing errors:

```
2 root error(s) found.
  (0) INVALID_ARGUMENT:  You must feed a value for placeholder tensor 'cond/else/_1/cond/Placeholder' with dtype int32
         [[{{node cond/Placeholder}}]]
         [[Func/cond/then/_0/input/_29/_16]]
  (1) INVALID_ARGUMENT:  You must feed a value for placeholder tensor 'cond/else/_1/cond/Placeholder' with dtype int32
         [[{{node cond/Placeholder}}]]
```

This is the same error as in
- #19246 which seems to have been solved since from the discussions, but used the built-in training loop with stateless optimizers
- #20164 which used a custom training loop and a stateful optimizer and still fails, and for which I found that using a stateless optimizer or applying my proposed fix below works

## Reproduction
On latest master or v3.14.1 + TF 2.21.0, but I've seen it with older versions as well.
A logical device distributed context is enough to reproduce, no need for actual GPUs.
A very simple model is also enough.
```py
import keras
import tensorflow as tf

tf.config.set_logical_device_configuration(
    tf.config.list_physical_devices('CPU')[0],
    [tf.config.LogicalDeviceConfiguration()]*2
)

class Model(keras.Model):
    def __init__(self):
        super().__init__()
        self.w = self.add_weight()
    def call(self, x):
        return self.w * x

# cross_device_ops=tf.distribute.ReductionToOneDevice() works
strategy = tf.distribute.MirroredStrategy()
with strategy.scope():
    inputs, targets = tf.random.normal((1,)), tf.random.normal((1,))
    opt = keras.optimizers.SGD(momentum=.1) # momentum=0. works
    model = Model()

@tf.function # removing decorator works
def replica_fn(inputs, targets):
    with tf.GradientTape() as tape:
        pred = model(inputs)
        loss = keras.losses.mean_absolute_error(pred, targets)
    grads = tape.gradient(loss, model.trainable_variables)
    opt.apply_gradients(zip(grads, model.trainable_variables))

strategy.run(replica_fn, args=(inputs, targets))

```

## Workarounds
- Set `cross_device_ops=tf.distribute.ReductionToOneDevice()` in `MirroredStrategy`
- Use state/variable-less optimizers, e.g. SGD with no momentum
- Use a single device
- Remove `@tf.function`

## Suggested fix
While going through the optimizers / TF backend code with the help of an AI assistant, it proposed the use of [`tf.init_scope`](https://www.tensorflow.org/api_docs/python/tf/init_scope) when creating optimizer variables.
Wrapping `add_variable` within a `tf.init_scope` context in `TFOptimizer` indeed solves my issue and seems to keep optimizer tests OK.
Something like:
```py
def add_variable(self, *args, **kwargs):
    with tf.init_scope():
        return super().add_variable(*args, **kwargs)

```

My understanding is that `tf.init_scope` is made for this type of issue 
> There is often a need to lift variable initialization ops out of control-flow scopes, function-building graphs, and gradient tapes.

but I must admit I'm not 100% familiar with it to see if there could be any unintended side effects.

If you think this solution is sound, I'd be happy to open a PR for it.
Otherwise let me know if there's anything I can do to help.

Thank you

**Comment by maitry63** [comment]
Hi @peinguim , Thank you for reporting this & your interest!

Whenever you create a variable inside a compiled graph like `@tf.function`, TensorFlow gets stuck trying to build static placeholders for a dynamic operation, which causes that `placeholder error`. To prevent this, we have `tf.init_scope()` as an explicit escape that lifts variable creation out of the active graph and safely pushes it into the global scope where variables belong.

Also , Keras 3 is backend agnostic , it should work seamlessly across JAX, PyTorch, and TensorFlow, the whole variable system is written.But JAX and PyTorch handle code and variables very differently than TensorFlow.

So, your suggested fix is correct, please go ahead and open a PR.

**Comment by peinguim** [comment]
Thank you so much @maitry63! I've opened #23087 with the fix
